scalableminds/webknossos
# Screenshot test tracing view

This PR adds screenshot tests using puppeteer. Since puppeteer can render webGL now, we're able to open the tracing view and screenshot the canvas. This allows us to do pixel perfect comparison tests of our rendering output.
Currently, I've added a first simple screenshot test. The 6 datasets that are available on our dev instances are opened at their default position/zoomstep, then screenshot and compared to the "ground truth" screenshots that are checked into the repository.

The tests are not ran as part of our CI, but instead can be triggered on demand by executing the `docker-compose up screenshot-tests` command. This will use the instance at `master.webknossos.xyz` by default. A specific dev instance can be used by specifying the url as follows: `URL=puppeteertracingview.webknossos.xyz docker-compose up screenshot-tests`.
To renew the screenshots, simply delete them, they are then newly created when running the tests for the next time.

#### Missing

  - [x] In order for puppeteer to authenticate itself, it would be helpful if on the dev instances a static authentication token could be inserted into the database (as is being done with the scmboy user). We could then hardcode this authentication token in the screenshot tests. @rschwanhold or @youri-k  Do you think this is possible?
  - [x] I'm not sure whether the Dockerfile.puppeteer is in the correct place here, maybe it should go into our Dockerfile repository? @jstriebel (Could you also review the changes I made to the docker-compose.yml please :) )
  - [x] Remove Dockerfile.puppeteer once, https://github.com/scalableminds/dockerfiles/pull/13 is merged.


### Steps to test:
- So far I've only tested this on my machine, so I'm not sure what the chances are that this works out-of-the-box for everyone, let's see :joy: 
- Checkout the branch and run `docker-compose up screenshot-tests`. Ideally the tests should be successful :pray:

### Issues:
- fixes #2822

------
- [x] Updated [changelog](https://github.com/scalableminds/webknossos/blob/master/CHANGELOG.md#unreleased)
- [x] Updated [migration guide](https://github.com/scalableminds/webknossos/blob/master/MIGRATIONS.md#unreleased) if applicable
